[Xcode] Clean up OTHER_LDFLAGS for WebKit services

ddkilzer opened the pull request:
#### 72c0e09bb8dc2b86ffa8c2527398b64b71649f21
<pre>
[Xcode] Clean up OTHER_LDFLAGS for WebKit services
<a href="https://bugs.webkit.org/show_bug.cgi?id=252220">https://bugs.webkit.org/show_bug.cgi?id=252220</a>
&lt;<a href="https://rdar.apple.com/105427340">rdar://105427340</a>&gt;

Reviewed by Elliott Williams.

Each XPC service xcconfig duplicated flags already defined in
BaseXPCService.xcconfig (such as `$(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH)`
and `$(WK_RELOCATABLE_FRAMEWORKS_LDFLAGS)`) and independently added
`$(SOURCE_VERSION_LDFLAGS)`.  Centralize `$(SOURCE_VERSION_LDFLAGS)` in
BaseXPCService.xcconfig and remove the duplicated flags from each
service, keeping only service-specific additions via `$(inherited)`.

Use the `:default=` operator so `OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH`
in BaseTarget.xcconfig reads `$(WK_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG)`
when set, falling back to `$(USE_STAGING_INSTALL_PATH)`.  This lets
WebContentService.xcconfig override the staged-framework-path condition
by setting `WK_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG` rather than
duplicating the entire `OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH`
computation.

* Source/WebKit/Configurations/BaseTarget.xcconfig:
(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH):
- Use `:default=` to read `WK_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG`
  with `$(USE_STAGING_INSTALL_PATH)` as fallback.
* Source/WebKit/Configurations/BaseXPCService.xcconfig:
(OTHER_LDFLAGS):
- Add `$(SOURCE_VERSION_LDFLAGS)`.
* Source/WebKit/Configurations/GPUService.xcconfig:
(OTHER_LDFLAGS): Delete.
* Source/WebKit/Configurations/NetworkService.xcconfig:
(OTHER_LDFLAGS): Delete.
(OTHER_LDFLAGS[sdk=macosx*]): Delete.
(ENTITLEMENTS_REQUIRED):
- Add missing semicolon.
* Source/WebKit/Configurations/WebContentService.xcconfig:
(WK_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG):
- Set from `$(WK_WEBCONTENT_SERVICE_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG)`
  to override the BaseTarget.xcconfig default.
(OTHER_LDFLAGS):
- Keep only `$(WK_APPKIT_LDFLAGS)` and `$(OTHER_LDFLAGS_PLATFORM)`.

Canonical link: <a href="https://commits.webkit.org/310771@main">https://commits.webkit.org/310771@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c087f34330efebcbb1ecaf372fe89574c0ef827e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/154883 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/28142 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/21302 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/163643 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/108353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/3d4635b0-cd92-4bd0-97d9-b999d0b55d2d) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/28281 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/27991 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/119822 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/108353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/263dab85-26ce-4b19-a258-53fec36a7846) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/157842 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/22093 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/139091 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/100515 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/21178 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/19207 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/11469 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/130840 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/166117 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/9494 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/18544 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/127924 "Passed tests") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/27687 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/23247 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/128063 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/27611 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/138728 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/84316 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/23622 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/22941 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/15523 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/27303 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/91405 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/26881 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/27112 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/26954 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKit/Configurations/NetworkService.xcconfig:
```diff
@@ -28,10 +28,11 @@ PRODUCT_BUNDLE_IDENTIFIER = $(PRODUCT_NAME);
-OTHER_LDFLAGS[sdk=macosx*] = $(inherited) $(SOURCE_VERSION_LDFLAGS);
+// Override values in BaseTarget.xcconfig to disable PGO profile generation.
+PROFILE_GENERATE_OR_USE_CFLAGS_ENABLE_LLVM_PROFILE_GENERATION[sdk=macosx*] = ;
+PROFILE_GENERATE_OR_USE_LDFLAGS[sdk=macosx*] = ;
```
It feels wrong that we are disabling PGO _generation_ but not _usage_ (i.e. we aren't stubbing out all of PROFILE_GENERATE_OR_USE_CFLAGS, only the enable-generation branch). This means we are building the XPC services with PGO profiles that are based on instrumented framework code. It's probably not causing a performance loss, but it's unnecessary.

Similarly it feels wrong that PGO is completely enabled in GPUService.xcconfig.

I know this is not your bug, but since you're refactoring this, could you fix these issues and move all the PGO-disabling logic to BaseXPCService.xcconfig?

ddkilzer replied:
Yep, I can do this, but I have other tasks that are more important, so I don't have time to spend on this now.  (This was something I started to fix as part of <https://github.com/WebKit/WebKit/pull/3811>, but decided to split it out into its own PR.)

Will come back to it later.

ddkilzer replied:
Commit 279215@main (fd4a28e3df08be757551476a5a9f045b63d95367) via PR https://github.com/WebKit/WebKit/pull/59585 rendered this moot since all those variables have changed.  :)

ddkilzer commented on Source/WebKit/Configurations/WebContentService.xcconfig:
```diff
@@ -46,7 +46,11 @@ CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING = NO; // Disable PGO profile generat
-OTHER_LDFLAGS = $(inherited) $(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH_$(WK_WEBCONTENT_SERVICE_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG)) $(WK_APPKIT_LDFLAGS) $(OTHER_LDFLAGS_PLATFORM) $(WK_RELOCATABLE_FRAMEWORKS_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
+
+// Override value in BaseTarget.xcconfig.
+OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH = $(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH_$(WK_WEBCONTENT_SERVICE_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG)$(WK_XCODE_15_3));
```
Review this carefully; appears to be a regression from [279215@main](https://github.com/WebKit/WebKit/commit/b3e2d8b3c17e8cd046af3749e1bdc49291e9acbe) (PR https://github.com/WebKit/WebKit/pull/28938).

aproskuryakov commented on Source/WebKit/Configurations/WebContentService.xcconfig:
```diff
@@ -46,7 +46,11 @@ CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING = NO; // Disable PGO profile generat
 WK_PGO_GEN_FLAGS = ; // Disable IR PGO profile generation
 WK_PGO_USE_CFLAGS = ; // Disable IR PGO profile use
 WK_PGO_USE_LDFLAGS = ; // Disable IR PGO profile use
-OTHER_LDFLAGS = $(inherited) $(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH_$(WK_WEBCONTENT_SERVICE_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG)) $(WK_APPKIT_LDFLAGS) $(OTHER_LDFLAGS_PLATFORM) $(WK_RELOCATABLE_FRAMEWORKS_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
```
I'm confused about the premise of this PR. `OTHER_LDFLAGS` includes `$(inherited)` here, so it already extends the base one, is that not so?

ddkilzer replied:
It started out as a clean-up patch to remove duplication in `OTHER_LDFLAGS`.

aproskuryakov replied:
I'd prefer to rely on `$(inherited)` and not add `WK_SERVICE_EXTRA_OTHER_LDFLAGS`, as it's unusual and thus confusing.

ddkilzer replied:
Fixed!

aproskuryakov commented on Source/WebKit/Configurations/WebContentService.xcconfig:
```diff
@@ -46,7 +46,11 @@ CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING = NO; // Disable PGO profile generat
-OTHER_LDFLAGS = $(inherited) $(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH_$(WK_WEBCONTENT_SERVICE_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG)) $(WK_APPKIT_LDFLAGS) $(OTHER_LDFLAGS_PLATFORM) $(WK_RELOCATABLE_FRAMEWORKS_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
+
+// Override value in BaseTarget.xcconfig.
+OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH = $(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH_$(WK_WEBCONTENT_SERVICE_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG)$(WK_XCODE_15_3));
```
We don't do Xcode 15.3 any more, the minimum is 26.2.

ddkilzer replied:
Good to know.  Since `$(WK_XCODE_15_3)` was still being used in the tree, this change was to align with that.

ddkilzer replied:
I think we should remove `$(WK_XCODE_15_3)` in a separate PR.  (Although I could be convinced to combine it if that's what folks want.)

aproskuryakov reviewed (commented):
I think that the commit message is still incorrect:

> Each XPC service xcconfig overrode OTHER_LDFLAGS entirely instead
> of extending BaseXPCService.xcconfig's definition

aproskuryakov commented on Source/WebKit/Configurations/BaseTarget.xcconfig:
```diff
@@ -77,6 +77,7 @@ OTHER_LDFLAGS = $(inherited) $(LLVM_PROFILE_GENERATION_OTHER_LD_FLAGS) $(WK_PGO_
 
 // FIXME: This shouldn't be computing path from scratch, the computation should be shared.
+// Overridden in WebContentService.xcconfig.
 OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH = $(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH_$(USE_STAGING_INSTALL_PATH)$(WK_XCODE_15_3));
```
I'm not sure if WebContentService.xcconfig would override the use in BaseTarget.xcconfig‎, I think that xcconfig includes aren't textual like C ones.

The PR looks good to me, but due to this detail, I'd like to leave it for @emw-apple to review.

emw-apple replied:
Yeah, this will work. When a build setting is defined more than once in a settings layer, the second definition (in top-to-bottom include order) wins.

emw-apple commented on Source/WebKit/Configurations/BaseTarget.xcconfig:
```diff
@@ -77,6 +77,7 @@ OTHER_LDFLAGS = $(inherited) $(LLVM_PROFILE_GENERATION_OTHER_LD_FLAGS) $(WK_PGO_
 
 // FIXME: This shouldn't be computing path from scratch, the computation should be shared.
+// Overridden in WebContentService.xcconfig.
 OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH = $(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH_$(USE_STAGING_INSTALL_PATH)$(WK_XCODE_15_3));
```
I think it would be clearer to:
- Rename WK_WEBCONTENT_SERVICE_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG (in WebContentService) to something more generic, like `WK_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG`
- Read that setting here directly, like:

```suggestion
OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH = $(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH_$(WK_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG:default=$(USE_STAGING_INSTALL_PATH))$(WK_XCODE_15_3));
```

Build settings are lazily evaluated, so WK_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG being defined "below" this line does not change how it expands.

(also, `:default=` is a quicker way to provide a default value than parametrizing using `$(WK_EMPTY_...)` or similar)

emw-apple reviewed (approved):
Makes sense, thank you!

Commit:
[Xcode] Clean up OTHER_LDFLAGS for WebKit services
https://bugs.webkit.org/show_bug.cgi?id=252220
<rdar://105427340>

Reviewed by Elliott Williams.

Each XPC service xcconfig duplicated flags already defined in
BaseXPCService.xcconfig (such as `$(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH)`
and `$(WK_RELOCATABLE_FRAMEWORKS_LDFLAGS)`) and independently added
`$(SOURCE_VERSION_LDFLAGS)`.  Centralize `$(SOURCE_VERSION_LDFLAGS)` in
BaseXPCService.xcconfig and remove the duplicated flags from each
service, keeping only service-specific additions via `$(inherited)`.

Use the `:default=` operator so `OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH`
in BaseTarget.xcconfig reads `$(WK_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG)`
when set, falling back to `$(USE_STAGING_INSTALL_PATH)`.  This lets
WebContentService.xcconfig override the staged-framework-path condition
by setting `WK_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG` rather than
duplicating the entire `OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH`
computation.

* Source/WebKit/Configurations/BaseTarget.xcconfig:
(OTHER_LDFLAGS_STAGED_FRAMEWORK_PATH):
- Use `:default=` to read `WK_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG`
  with `$(USE_STAGING_INSTALL_PATH)` as fallback.
* Source/WebKit/Configurations/BaseXPCService.xcconfig:
(OTHER_LDFLAGS):
- Add `$(SOURCE_VERSION_LDFLAGS)`.
* Source/WebKit/Configurations/GPUService.xcconfig:
(OTHER_LDFLAGS): Delete.
* Source/WebKit/Configurations/NetworkService.xcconfig:
(OTHER_LDFLAGS): Delete.
(OTHER_LDFLAGS[sdk=macosx*]): Delete.
(ENTITLEMENTS_REQUIRED):
- Add missing semicolon.
* Source/WebKit/Configurations/WebContentService.xcconfig:
(WK_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG):
- Set from `$(WK_WEBCONTENT_SERVICE_NEEDS_STAGED_FRAMEWORK_PATH_LDFLAG)`
  to override the BaseTarget.xcconfig default.
(OTHER_LDFLAGS):
- Keep only `$(WK_APPKIT_LDFLAGS)` and `$(OTHER_LDFLAGS_PLATFORM)`.

Canonical link: https://commits.webkit.org/310771@main
